### DISCUSSION TITLE
usage/cli/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#16701] [Docs]
# usage/cli/

Explore the YOLO11 command line interface (CLI) for easy execution of detection tasks without needing a Python environment.

https://docs.ultralytics.com/usage/cli/

**Comment by ashikrasul** [comment]
Is there a way to pass the hypeparameter file in CLI during training?

**Reply by glenn-jocher** [reply]
@arasul42 yes, you can pass a hyperparameter file in the CLI using the `cfg` argument, like this: `yolo train cfg=your_hyperparameters.yaml`.

**Reply by muhammadrizwan11** [reply]
thank you so much i was confused but you give us a good thing

**Reply by glenn-jocher** [reply]
You're welcome! 😊 We're glad the CLI guidance helped clarify things. The Ultralytics team and community are always here if you have more questions – just drop them in GitHub Discussions. Happy training with YOLO!

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for engaging with the Ultralytics community 🚀!

It looks like you're interested in exploring the YOLO11 command line interface (CLI). We recommend checking out our [CLI Usage Documentation](https://docs.ultralytics.com/usage/cli/) for comprehensive guidance and examples that can help you get started.

If this is a 🐛 Bug Report related to CLI usage, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to aid in debugging.

If you have any questions or need assistance with custom training ❓, ensure you include detailed information like command outputs and configurations. You can also review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our community on [Discord](https://ultralytics.com/discord) for real-time support 🎧, check out our [Discourse](https://community.ultralytics.com) for detailed discussions, or connect on our [Subreddit](https://reddit.com/r/ultralytics) to exchange ideas with other enthusiasts.

## Upgrade

Ensure you are using the latest version by upgrading the `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) to avoid any outdated issues:

```bash
pip install -U ultralytics
```

## Environments

To run YOLO in well-supported environments, consider these options:

- **Notebooks** with free GPU support: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud environments**: Refer to our quickstart guides for [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) and [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker**: Utilize our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This badge indicates if all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. CI tests ensure functionality across macOS, Windows, and Ubuntu and are run daily and on every commit.

This is an automated response, but an Ultralytics engineer will join the conversation shortly to provide more insights.

**Comment by sakrig** [comment]
Apple's NeuralEngine API accelerates models for all new Apple products with the M4 chips.

Does the IOS CLI app use the NeuralEngine for all Apple M4 products?
Does Ultralytics support the M4 processor on the latest Mac, MacMini, Mac Pro, iPad, iPhone ...?

Seems straight forward to support the M4 NeuralEngine for all Apple products, but I have not tried.

**Reply by glenn-jocher** [reply]
@sakrig ultralytics YOLO models can be exported to CoreML format for Apple devices, which leverages Apple's native acceleration including NeuralEngine where available. For M4 support, ensure you're using the latest PyTorch versions with macOS MPS backend. See our [CoreML export docs](https://docs.ultralytics.com/modes/export/#coreml) for implementation details.

**Comment by aryakpt** [comment]
HI, can you help me, how do i save the validation result if i use CLI? Because i clone github ultralytics directly

**Reply by glenn-jocher** [reply]
@aryakpt to save validation results when using the CLI, add `save_json=True` to your validation command. For example:  
`yolo val model=path/to/

**Reply by aryakpt** [reply]
thanks glenn for the response. but how can i access the val result like this using CLI? I want to save it in a variable

```
best_model = YOLO(f"{output_path}/train{suffix}/weights/best.pt")
val_result = best_model.val(data=data,  project=output_path)
map50 = val_result.box.map50
```

**Reply by glenn-jocher** [reply]
To capture validation metrics programmatically via CLI, use the `save_json=True` argument which saves results to a JSON file. You can then parse this JSON output to access metrics like mAP50. The CLI doesn't directly return variables, but the JSON file provides structured results for programmatic access. For details, see the Val mode documentation on saving validation outputs.

**Comment by muhammadrizwan11** [comment]
I noticed some discrepancies in the training and validation commands under the CLI chapter of the Ultralytics YOLO documentation. The documented training command is written as '' yolo detect train data=coco8.yaml model=yolo11n.pt epochs=100 imgsz=640 '' but the correct command should be '' yolo train data=coco8.yaml model=yolo11n.pt epochs=10 lr0=0.01'' The word "detect" should be removed. Similarly, the validation command is written as '' yolo detect val model=yolo11n.pt'' , whereas the correct command should be '' yolo val model=yolo11n.pt data=coco8.yaml batch=1 imgsz=640 ''. Again, "detect" is unnecessary, and important parameters such as the dataset (data=coco8.yaml), batch size (batch=1), and image size (imgsz=640) are missing. These inconsistencies might cause confusion for users following the documentation. Could the team review and update this section for accuracy? 
i am also doing working in yolo https://github.com/muhammadrizwan11

**Reply by glenn-jocher** [reply]
@muhammadrizwan11 thank you for pointing this out! You're correct that the `detect` task argument is optional in CLI commands when the model type is specified, as YOLO can infer the task from the model name. Both `yolo detect train` and `yolo train` syntaxes are valid, but we'll standardize the docs to use the more concise form where possible. The validation example should indeed include explicit parameters for clarity, even though `data` is inherited from training. We'll update the documentation to reflect these points. For reference, see the CLI Syntax section in the Ultralytics YOLO documentation.

**Comment by MUK003** [comment]
Hello Team Ultralytics,

Good morning!

Many thanks for the great work.

Couple of questions:
1. How can i generate same graphs (graphs after training) after using the trained model on my unsense samples data?
2. I can use the target folder after training the model, however, it only exports the images with the prediction boxes. I want to have a pie chart or bar chart or even the same plots it exports after the training run
3. For the object detection via a video, if i have 4 images to display in one second, the model looks the same image multiple times. How can we correct the frames reading?


Looking forward to your kind input,

Kind regards,

MUK
